src/artemis/devices/fast_grid_scan.py

```diff
@@ -0,0 +1,218 @@
+import threading
+import time
+from typing import List
+from ophyd import (
+    Component,
+    Device,
+    EpicsSignal,
+    EpicsSignalRO,
+    EpicsSignalWithRBV,
+    Signal,
+)
+from ophyd.status import DeviceStatus, StatusBase, SubscriptionStatus
+
+from dataclasses import dataclass
+from typing import Any
+
+from src.artemis.devices.motors import (
+    GridScanLimit,
+    GridScanLimitBundle,
+)
+
+from bluesky.plan_stubs import mv
+
+
+@dataclass
+class GridScanParams:
+    """
+    Holder class for the parameters of a grid scan.
+    """
+
+    x_steps: int = 1
+    y_steps: int = 1
+    x_step_size: float = 0.1
+    y_step_size: float = 0.1
+    dwell_time: float = 0.1
+    x_start: float = 0.1
+    y1_start: float = 0.1
+    z1_start: float = 0.1
+
+    def is_valid(self, limits: GridScanLimitBundle) -> bool:
+        """
+        Validates scan parameters
+
+        :param limits: The motor limits against which to validate
+                       the parameters
+        :return: True if the scan is valid
+        """
+
+        return (
+            # All scan axes are within limits
+            scan_in_limits(limits.x, self.x_start, self.x_steps, self.x_step_size)
+            and scan_in_limits(limits.y, self.y1_start, self.y_steps, self.y_step_size)
+            # Z never exceeds limits
+            and limits.z.is_within(self.z1_start)
+        )
+
+
+def scan_in_limits(
+    limit: GridScanLimit, start: float, steps: float, step_size: float
+) -> bool:
+    end = start + (steps * step_size)
+    return limit.is_within(start) and limit.is_within(end)
+
+
+class GridScanCompleteStatus(DeviceStatus):
+    """
+    A Status for the grid scan completion
+    A special status object that notifies watchers (progress bars)
+    based on comparing device.expected_images to device.position_counter.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.start_ts = time.time()
+
+        self.device.position_counter.subscribe(self._notify_watchers)
+        self.device.status.subscribe(self._running_changed)
+
+        self._name = self.device.name
+        self._target_count = self.device.expected_images.get()
+
+    def _notify_watchers(self, value, *args, **kwargs):
+        if not self._watchers:
+            return
+        time_elapsed = time.time() - self.start_ts
+        try:
+            fraction = value / self._target_count
+        except ZeroDivisionError:
+            fraction = 1
+            time_remaining = 0
+        except Exception:
```

Comment:
If there is some unknown exception I think we should pass it upwards with `self.set_exception()`.